src/Features/LanguageServer/Protocol/Handler/Diagnostics/DocumentPullDiagnosticHandler.cs

```diff
@@ -0,0 +1,154 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics
+{
+    using LspDiagnostic = Microsoft.VisualStudio.LanguageServer.Protocol.Diagnostic;
+
+    [ExportLspMethod(MSLSPMethods.DocumentPullDiagnosticName), Shared]
+    internal class DocumentPullDiagnosticHandler : AbstractRequestHandler<DocumentDiagnosticsParams, DiagnosticReport[]?>
+    {
+        private static readonly DiagnosticTag[] s_unnecessaryTags = new[] { DiagnosticTag.Unnecessary };
+
+        private readonly IDiagnosticService _diagnosticService;
+
+        /// <summary>
+        /// Lock to product <see cref="_documentIdToLastResultId"/> and <see cref="_nextResultId"/>.
+        /// </summary>
+        private readonly object _gate = new object();
+
+        /// <summary>
+        /// Mapping of a document to the last result id we reported for it.
+        /// </summary>
+        private readonly Dictionary<(Workspace workspace, DocumentId documentId), string> _documentIdToLastResultId =
+            new Dictionary<(Workspace workspace, DocumentId documentId), string>();
+
+        /// <summary>
+        /// The next available id to label results with.
+        /// </summary>
+        private long _nextResultId;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public DocumentPullDiagnosticHandler(
+            ILspSolutionProvider solutionProvider,
+            IDiagnosticService diagnosticService) : base(solutionProvider)
+        {
+            _diagnosticService = diagnosticService;
+            _diagnosticService.DiagnosticsUpdated += OnDiagnosticsUpdated;
+        }
+
+        private void OnDiagnosticsUpdated(object? sender, DiagnosticsUpdatedArgs updateArgs)
```

Comment:
Why do we need to listen to diagnostic updated events?
Is this to avoid calling GetDiagnostics when nothing has changed since the last LSP pull request?  If so, could we instead just set a bool that says if we have new diagnostics, or do we need to know the exact diagnostics we last sent

Reply:
>  or do we need to know the exact diagnostics we last sent

this is teh primary problem.  the current approach means we don't need to keep track of which diagnostics were sent (or do any sort of attempt at comparing current/previous diagnostics).  All we do is just mark if the diagnostics for a doc were already sent out.

when things then change, we just 'clear/forget' that the client knows about them, so we return them in the future when asked.